man/mkdotenv.1
.TH MKDOTENV 1 "July 2026" "mkdotenv 1.0.1"
.SH NAME
mkdotenv \- A command-line tool that populates secrets upon a .env file from a template.
.SH DESCRIPTION
The \fmkdotenv\fR command allows users to poopulate environmental variables from a template file by placing appropriate markup upon the file.
.SH SYNOPSIS
.B mkdotenv
[\fI--help\fR| \fI-h\fR][\fI--version\fR| \fI-v\fR]
[\fI--environment\fR| \fI-e\fR | \fI--env\fR <\fIENVIRONMENT\fR>]
[\fI--template-file\fR| \fI-t\fR | \fI--template\fR <\fITEMPLATE-FILE\fR>][\fI--output-file\fR| \fI-o\fR <\fIOUTPUT-FILE\fR>]
[\fI--argument\fR| \fI-a\fR | \fI--arg\fR <\fIARGUMENT\fR>]

.SH Optional ARGUMENTS

.TP
.B --help,\fI-h\fR
Display help message and exit
.TP
.B --version,\fI-v\fR
Display version and exit
.TP
.B --environment,\fI-e\fR ,\fI--env\fR
Environment in which secrets would be resolved
.TP
.B --template-file,\fI-t\fR ,\fI--template\fR
Template .env file containing commands on how .env file would be generated. If no vanue provided .env.dist assumed.
.TP
.B --output-file,\fI-o\fR
File to write output to
.TP
.B --argument,\fI-a\fR ,\fI--arg\fR
Argument provided as $_ARG upon template file.
.SH AUTHOR
Written by Desyllas Dimitrios.

.SH BUGS
Report issues at https://github.com/pc-magas/mkdotenv/issues

.SH SEE ALSO
.BR dotenv (1)


Changelog.md
# Version 1.0.1 2026-02-12

Fix PPA build.

# Version 1.0.0 2026-01-05

1. Introduce secret resolvers
2. [Breaking Change] Restructure cli command line arguments
3. [Breaking change] Use template file and a Description Language on it in order to resolve secrets.
4. Tool to generate manpage from Arguments

# Version 0.4.9 2025-09-14

Enable GCO on some architectures, fine tuning unmapped architectures as well upon MAKEFILE.

# Version 0.4.8 2025-09-13

Enable GCO on some architectures also allow default GCO architectures as well.

# Version 0.4.7 2025-09-12

Autoenable gco support on some architectures. Windows only for x86_64 only.

# Version 0.4.6 2025-09-12

Add extra supported values for building on alpine aports ci/cd.

# Version 0.4.5 2025-09-12

Vendor dependencies upon alpine vendor source.

# Version 0.4.4 2025-09-12

Alpine changes upon Makefile.
Autogenerate APKBUILD for alpine builds.

# Version 0.4.3 2025-09-03

Fix version upon manpage and remove non-existent arguments
Fix pipeline upon releasing macos homebrew fromula.

# Version 0.4.2 2025-08-27

Use Seperate script for debian builds
Vendor dependencies upon build.
Introducing tools for PPA and debian building
Improve Documentation
Use `pflag` for argument parsing

# Version 0.4.1 2025-08-20

Use Seperate script for debian builds

# Version 0.4.0 2025-07-18

Fix alpine build.
Use variable-value parameter for setting the value as variable value.
Use variable-name parameter for setting the value as variable value.
[NEW FEATURE] Flag to remove multiple occurences of the variable.
[NEW FEATURE] Use - value upon in order to output modified .env contents upon stdout. Default behaviour is outputing upon .env

# Version 0.3.4 2025-07-01

Add release number upon ppa build.
Fix alpine build

# Version 0.3.3 2025-06-26

Improved README (fixing typos and improve clarification).
Arch package does installs oficial golang and does not requires it as dependency.
Using ./fedora folder for rpm building.
Improvement upon debian building
Fixing ppa release

# Version 0.3.2 2025-06-12

Release for MACOS
Use ghcr hosted images for fedora and alpine builds
Native building of app usiong Make both on MacOs and Linux.

# Version 0.3.1 2025-05-26

1. [BUGFIX] Check Variable name
2. Use External Docker image for rpm builds.

# Version 0.3.0 2025-05-12

1. Use common naming convention for golang module using repoesitory's name
2. Upon rpm builds use Makefile
3. Ability to specify a version externally in Makefile.
4. Unit test value appending logic
5. Validate variable name
6. Moving pcmagas/alpinebuild (used upon alpine image releases) docker image into a seperate repository.
7. Release for AUR and arch linux

# Version 0.2.3 2025-04-24

Release for Alpine

# Version 0.2.2 2025-04-10

Release for windows
make runs `bin` target by default
Fix lang upon rpm changelog

# Version 0.2.1 2025-04-07

1. Improve Argument parsing

# Version 0.2.0 2025-03-27

1. Split codebase into multiple files.
2. Use a seperate version file and define built version upon compile.
4. [BUGFIX] If input file is same as output file copy input file into a temporary one.
5. Improved Documentation
6. [BUGFIX] Out of bounds argument parsing
7. [BUGFIX] Values should not be an Argument

# Version 0.1.7 2025-02-23

1. Release rpm image

# Version 0.1.5 2025-02-12

1. Build for PPA

# Version 0.1.0 2025-02-02

1. Build Debian package
2. Add build from sources upon documentation
3. Build Docker image
4. If no arguments provided print usage.

# Version 0.0.1 2025-01-31

1. Basic .env manipulation
2. Write .env files
3. Print .env to stdout
4. Receive .env contents from unix pipes.


src/mkdotenv.go
/**
*   MkDotenv the .env file manipulator
*    Copyright (C) 2024 Desyllas Dimitrios
*
*   This program is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package main

import (
    "os"
    "fmt"
    "bufio"
	"strings"
	"regexp"
	"errors"
	"slices"
)

const VERSION = "0.1.7"

func printHelp() {
	printVersion()

	fmt.Println("\nUsage:\n\t"+os.Args[0]+" <variable_name> <variable_value> [--env-file | --input-file <file_path>] [--output-file <file_path>]\n")
	fmt.Println("Arguments:")
	fmt.Println("\tvariable_name\tREQUIRED The name of the variable")
	fmt.Println("\tvariable_value\tREQUIRED The value of the variable prtovided upon <variable_name>")
	fmt.Println("\nOptions:")
	fmt.Println("\t--env-file (or --input-file) <file_path> \tOPTIONAL The .env file path in <file_path> that will be manipulated. Default value .env")
	fmt.Println("\t--output-file <file_path> \tOPTIONAL Instead of printing the result into console write it into a file. If value provided it will NOT output the contents of the .env file.")
}

func printVersion(){
	fmt.Println("\nMkDotenv VERSION: ",VERSION)
	fmt.Println("Replace or add a variable into a .env file.")
}


func append_value_to_dotenv(file *os.File,output *bufio.Writer,variable_name string,variable_value string) (bool,error) {
	
	var newline string = fmt.Sprintf("%s=%s", variable_name, variable_value)

	// If no .env exists then output the variable.
	if (file == nil){
		output.WriteString(newline+"\n")
		return true,nil
	}

	scanner := bufio.NewScanner(file)

	var variableFound bool = false

	variable_name=strings.TrimSpace(variable_name)

	if(variable_name == ""){
		return false,errors.New("Variable name is empty")
	}

	re, err := regexp.Compile(`^#?\s*`+variable_name+`\s*=.*`)
	if err != nil {
		return false,err
	}

	
	for scanner.Scan() {
		line:=scanner.Text()
		line_to_write:=line

		if re.MatchString(line) {
			line_to_write = newline
			variableFound=true
		}
		output.WriteString(line_to_write+"\n")
	}

	if !variableFound {
		output.WriteString(newline+"\n")
	}

	return true,nil
}


func getParameters()(string,string,string,string){
    
	if(len(os.Args) < 3){
        fmt.Fprintln(os.Stderr,"Not enough arguments provided")
		printHelp()
		os.Exit(1)
    }

    var dotenv_filename string = ".env"
    var variable_name string = os.Args[1]
	var variable_value string = os.Args[2]
	var output_file string = ""

	if(strings.HasPrefix(variable_name,"-")){
		printHelp()
		fmt.Fprintln(os.Stderr,"Variable Name should not start with - or --")
		os.Exit(1)
	}

	ARGUMENTS:= []string{"--env-file","--input-file","--output-file","-v","--version","-h","--h","--help"}

	if(slices.Contains(ARGUMENTS[:],variable_value)){
		printHelp()
		fmt.Fprintln(os.Stderr,"\nVariable value should not contain any of the values:\n"+strings.Join(ARGUMENTS[:],"\n"))
		os.Exit(1)
	}

	for i, arg := range os.Args[3:] {

		switch arg {
		 	case "--input-file":
				fallthrough;
			case "--env-file":
				// Arguments are parsed with an offset we get the next item + offset
				dotenv_filename = os.Args[i+3+1]
				
			case "--output-file":
				output_file = os.Args[i+3+1]
		}
	}
	
	return dotenv_filename,output_file,variable_name,variable_value
}

func printVersionOrHelp(){

	if(len(os.Args) > 2 ){
		return
	}


	switch(os.Args[1]){
		case "-h":
			fallthrough
		case "--help":
			printHelp()
			os.Exit(0)
		case "-v":
			fallthrough
		case "--version":
			printVersion()
			os.Exit(0)
		default:
			fmt.Fprintln(os.Stderr,"Not enough arguments provided")
			printHelp()
			os.Exit(1)
	}
}

func HandleFileError(err error, filename string) {
	if os.IsNotExist(err) {
		fmt.Fprintf(os.Stderr, "Error: The file '%s' does not exist.\n", filename)
	} else if os.IsPermission(err) {
		fmt.Fprintf(os.Stderr, "Error: Permission denied for file '%s'.\n", filename)
	} else {
		fmt.Fprintf(os.Stderr, "Error: Failed to open file '%s': %v\n", filename, err)
	}

	os.Exit(1)
}

func getFileToRead(dotenv_filename string) *os.File {

	var file *os.File
	var err error

	stat, _ := os.Stdin.Stat()
	hasPipeInput := (stat.Mode() & os.ModeCharDevice) == 0

	if dotenv_filename != ".env" { 
		// User explicitly provided a file, use it
		file, err = os.Open(dotenv_filename)
		if err != nil {
			HandleFileError(err, dotenv_filename)
			os.Exit(1)
		}
	} else if hasPipeInput {
		// No --env-file, but we have pipe input
		file = os.Stdin
	} else {
		// Default to .env
		file, err = os.Open(".env")
		if err != nil {

			if(os.IsNotExist(err)){
				return nil
			}

			HandleFileError(err, ".env")
			os.Exit(1)
		}
	}

	return file
}

func main() {

	if (len(os.Args) == 1 ){
		printHelp()
		os.Exit(0)
	}

	printVersionOrHelp()

	dotenv_filename,output_file,variable_name,variable_value := getParameters()

	file:=getFileToRead(dotenv_filename)
	defer file.Close()

	writer := bufio.NewWriter(os.Stdout)
	if output_file != "" {
		outfile,err := os.OpenFile(output_file,os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			HandleFileError(err,output_file)
		}
		defer file.Close()
		writer = bufio.NewWriter(outfile)
	}
	defer writer.Flush()

    _,err := append_value_to_dotenv(file,writer,variable_name,variable_value)
    if(err!=nil){
        fmt.Fprintln(os.Stderr, "Error:", err)
        os.Exit(1)
    }
}